### ISSUE TITLE
Bulk API / converts records to CSV which prevents all properties from being set

### ISSUE THREAD

**Issue by dbhoot** [jsforce/jsforce#754]
I think I've found a bug in the bulk API which causes an issue where not all fields of a record are updated.

Suppose I want to update "SOME_TABLE" in salesforce using the upsert operation. I create an array of records:

[ { property1: 'w' }, { property1: 'x', property2: 'y' }, { propety1: 'z', property2: 'a', property3: 3 } ]

Notice how different records have different properties to be updated. When I call the bulk upsert function it converts data set into a CSV. The first record of the upsert only has a single property. So the csv ends up looking something like:

property1
'w'
'x'
'z'

As a result, property 2 on the second record and property 2 and property 3 on the third record never get updated.

**Comment by andymjames** [comment]
I was just burned by this unexpected behavior as well.

In searching to notice this issue, I also searched closed issues, and found this was originally reported in 2015 (https://github.com/jsforce/jsforce/issues/258) but dismissed as working as intended.

I can understand that resolving this has no simple solutions, but in that case it seems absolutely critical to **document this**. I've spent loads of time in the docs, and this is mentioned nowhere. And if it is, then you need to replicate it to the many different spots where bulk write operations are discussed.

I am actually pretty upset right now, to think the devs here have known they were silently ignoring data in some cases for over nearly 3 years, and couldn't be bothered to document it nor find a different solution that is robust against the issue.

**Comment by dland512** [comment]
Yeah, just lost several hours to this as well. Glad I figured it out, but it does need to be documented. Is the documentation part of the github project? If so I'll gladly submit a PR.

**Comment by matthewpoer** [comment]
Found this one as well. This ticket is the only documentation on the issue and honestly is pretty hard to find. 

In our experience, we were sending several objects in a single Bulk JSForce Salesforce Bulk API call. These objects came from different use cases and had different key sets. It appears that keys added to the second object were completely discarded if that key didn't exist also on the first object.

Our workaround will be to normalize our objects' keys before passing to JSForce. 

It occurred to me that JSForce should do this for us, i.e. normalizing keys, but if it did then you could potentially set fields to empty values (false, empty strings, zero, whatever) instead of allowing default values to be set. 

The real need then is probably just to document the behavior. 

The more extreme option might be that the bulk api functionality within JSForce watches for additional keys to be introduces and throws an error when that happens.

**Comment by matthewpoer** [comment]
The bare minimum would be to document it: https://github.com/jsforce/jsforce-website/pull/17

**Comment by techieshark** [comment]
Noting that jsforce v3 is coming out:
https://github.com/jsforce/jsforce/issues/1360#issuecomment-1836701005

I haven't seen documentation on what is planned for v3 but it appears the offending lines (I think??) appear unchanged so would likely still need either a fix, or a workaround and some documented warning:

https://github.com/jsforce/jsforce/blob/5369d1d3e24bba0e384027d309c441c9ac5fa956/src/api/bulk.ts#L634C3-L643

```
      if (Array.isArray(input)) {
        for (const record of input) {
          for (const key of Object.keys(record)) {
            if (typeof record[key] === 'boolean') {
              record[key] = String(record[key]);
            }
          }
          this.write(record);
        }
        this.end();
```

**Comment by cristiand391** [comment]
Bulk 2.0 docs mention that empty values are ignored:

> Empty field values are ignored when you update records. To set a field value to null, use a field value of #N/A.

https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/datafiles_prepare_csv.htm

thinking out loud... maybe jsforce could safely build the CSV header to include all props but could hit perf the bigger an array is (we need to get all keys before serialization starts).
